### PR TITLE
fix(init): limit the amount of custom types created concurrently (DT-1758)

### PR DESCRIPTION
**Pull request by lihbr**
## Context

Since https://github.com/prismicio/slice-machine/releases/tag/slice-machine-ui%401.17.0, Slice Machine init command downloads remote custom types during the init process so that users can work on them locally.

When users have many custom types (60+), the init command can fail to allocate memory as it was creating all custom types concurrently which is a costly operation.

Resolves: DT-1758

## The Solution

This PR limits the rate at which custom types are created by the init command so that no more than 8 custom types are being created at the same time.

## How to QA

1. Create a repository with many custom types (60+) or ask me for a clone of such a repository
2. Try running the current init command inside a project, it **should** fail
```bash
npx create-next-app@latest
npx @slicemachine/init@latest
```
3. Try running the alpha init command, it **should** succeed
```bash
npx create-next-app@latest
npx @slicemachine/init@alpha
```

## Impact / Dependencies

/

## Checklist before requesting a review
- [ ] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [ ] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.

### CODE REVIEW

**Comment by lihbr** [comment]
Side notes:

We've been fixing multiple load issues recently using `p-limit` and other strategies. I'm considering making it more global by updating the [`HookSystem`](https://github.com/prismicio/slice-machine/blob/54a538b74763294f74647bd8c4155f52897bdeac/packages/plugin-kit/src/lib/HookSystem.ts) class to support an optional map of options for hooks. This would allow us to say something like this:

```typescript
import { HookSystem } from "./lib/HookSystem";
import type { SliceMachineHooks } from "./types";

/**
 * @internal
 */
export const createSliceMachineHookSystem =
	(): HookSystem<SliceMachineHooks> => {
		return new HookSystem<SliceMachineHooks>({
			"slice:create": { concurrencyLimit: 8 },
		});
	};
```

What do you think? @angeloashmore, you might have an opinion on this 🤔

File: package.json [hunk 1]
```diff
@@ -75,4 +75,4 @@
     "e2e-projects/next-upgrade",
     "packages/*"
   ]
-}
+}
```
Comment (bapmrl) [inline package.json]: ❓ Why this change? Running `yarn install` after this pull request is merged might generate an unnecessary diff, no?

File: packages/init/src/SliceMachineInitProcess.ts [hunk 2]
```diff
@@ -1286,9 +1287,17 @@ ${chalk.cyan("?")} Your Prismic repository name`.replace("\n", ""),
...
 									await this.manager.customTypes.fetchRemoteCustomTypes();
 
+								const createCustomTypeLimit = pLimit(8);
+
+								let pulled = 0;
+								task.title = `Pulling existing types... (0/${remoteTypes.length})`;
 								await Promise.all(
 									remoteTypes.map(async (model) => {
-										await this.manager.customTypes.createCustomType({ model });
+										await createCustomTypeLimit(() =>
```
Comment (bapmrl) [inline packages/init/src/SliceMachineInitProcess.ts:1296]: 💡 Apparently, it could also be written:
```
await limit(this.manager.customTypes.createCustomType, { model })
```

It's apparently used to avoid creating unnecessary closures, but we don't need that here ^^

**Review by bapmrl — APPROVED** [review]
Excellent! I just have 2 questions.

**Comment by bapmrl** [comment]
> Side notes:
> 
> We've been fixing multiple load issues recently using `p-limit` and other strategies. I'm considering making it more global by updating the [`HookSystem`](https://github.com/prismicio/slice-machine/blob/54a538b74763294f74647bd8c4155f52897bdeac/packages/plugin-kit/src/lib/HookSystem.ts) class to support an optional map of options for hooks. This would allow us to say something like this:
> 
> ```ts
> import { HookSystem } from "./lib/HookSystem";
> import type { SliceMachineHooks } from "./types";
> 
> /**
>  * @internal
>  */
> export const createSliceMachineHookSystem =
> 	(): HookSystem<SliceMachineHooks> => {
> 		return new HookSystem<SliceMachineHooks>({
> 			"slice:create": { concurrencyLimit: 8 },
> 		});
> 	};
> ```
> 
> What do you think? @angeloashmore, you might have an opinion on this 🤔

I'm not super aware of this topic, but searching for `p-limit`, it seems our use of it isn't necessarily tied to calling hooks and can happen anywhere we await a promise?

File: package.json (hunk 1 above)
Comment (lihbr) [reply package.json]: Yeah, I'll revert it, it's ESLint adding one, Yarn removing it, driving me crazy

File: packages/init/src/SliceMachineInitProcess.ts (hunk 2 above)
Comment (lihbr) [reply packages/init/src/SliceMachineInitProcess.ts:1296]: I've read about it too! Looking at the README, this way feels like the default one though unless we have performance concerns here beyond that 🤔

**Comment by lihbr** [comment]
Agreed yes, just seems like recent issues were mostly revolving around hooks

**Comment by bapmrl** [comment]
LGTM.
